[Bug]: AssetTargetFallback now brings in new dependencies and is causing build failures - Add a means to revert to the 17.1 behavior

### NuGet Product Used

MSBuild.exe

### Product Version

Microsoft (R) Build Engine version 17.2.0-preview-22081-04+288ea72fc for .NET Framework

### Worked before?

Microsoft (R) Build Engine version 17.1.0+ae57d105c for .NET Framework

### Impact

I'm unable to use this version

### Repro Steps & Context

As of the latest version, it seems like a missing dependency group breaks restore.

When packing, I do see this warning:
```
WARNING: NU5128: Some target frameworks declared in the dependencies group of the nuspec and the lib/ref folder do not have exact matches in the other location. Consult the list of actions below:
- Add a dependency group for .NETStandard2.0 to the nuspec
```

However, previously the behavior of a missing dependency group matched that of an empty dependency group.

Repro materials: [Repro.zip](https://github.com/NuGet/Home/files/8005402/Repro.zip)

Repro steps:
1. msbuild /restore TestLib
2. nuget pack -OutputDirectory Packages TestPackageBroken.nuspec
3. msbuild /t:restore TestProjectBroken

A warning is emitted:
```
C:\Code\tmp\RestoreRepro\TestProjectBroken\TestProjectBroken.csproj : warning NU1701: Package 'Microsoft.AspNet.WebApi.Core 5.2.7' was restored using '.NETFramework,Version=v4.6.1, .NETFramework,Version=v4.6.2, .NETFramewo
rk,Version=v4.7, .NETFramework,Version=v4.7.1, .NETFramework,Version=v4.7.2, .NETFramework,Version=v4.8' instead of the project target framework '.NETStandard,Version=v2.0'. This package may not be fully compatible with your
 project.
```

It works if you do this however (the package has an empty dependency group):
1. msbuild /restore TestLib
2. nuget pack -OutputDirectory Packages TestPackageWorks.nuspec
3. msbuild /t:restore TestProjectWorks

### Verbose Logs

_No response_

In the case of `Microsoft.Bcl`, looking at the assets file after restore, the only asset it has selected is `lib/net45/_._`, which is NuGet convention for none. Asking for a design change request (DCR) to stop NU1701 warnings when the only assets selected are `_._` seems reasonable.

However, in the case of Microsoft.Bcl.Async, the assets selected are `lib/net40/Microsoft.Threading.Tasks.Extensions.Desktop.dll`, `lib/net40/Microsoft.Threading.Tasks.Extensions.dll`, and `lib/net40/Microsoft.Threading.Tasks.dll`. In other words, this package is not correctly authored to act as a polyfill. In fact, the way the package is authored, NuGet will instruct the .NET SDK to copy the assemblies to the bin folder, and the .NET SDK itself has logic not to. But the way the package is authored is instructing NuGet and the build system to copy the files to the bin directory.

Also note that the latest version of the package was published in 2014, about 2 years before .NET Core 1.0 was released, and AssetTargetFallback was released later than that.

> However, previously the behavior of a missing dependency group matched that of an empty dependency group. 

The previous behaviour was that a missing dependency group matched the next best dependency group. It's only when none of the dependency groups matched that no dependencies were used.

Consider for example a package with `lib/net40/my.dll` and `lib/net48/my.dll`. The package only contains one dependency group for `.NETFramework4.0` to add `Microsoft.BCL.Async`. A project targeting .NET Framework 4.8 that installs this package will get the `lib/net48/` assets, and since there's a missing dependency group for `.NETFramework4.8`, it will select the next best, `.NETFramework4.0`. Therefore, the project will get a transitive dependency on Microsoft.Bcl.Async because the package is not authored in the highest quality way.

Another example is a package that has `lib/netstandard2.0/my.dll` and `lib/net5.0/my.dll`, but only a .NETStandard2.0 dependency group. Similarly, net5 and net6 projects will get the dependencies, even though the files used from the package are not the .NET Standard files.

NU5128 was added 2 years ago (not in VS 17.2) to try to encourage package authors to improve their package quality, and reduce the risk of these types of experiences for package consumers.

The problem that https://github.com/NuGet/Home/issues/5957 fixes is for scenarios along the lines of a `net6.0` project referencing a package that only supports `net48`, and it works via AssetTargetFallback, but fails at runtime because the package's dependencies are missing.

As the design spec for respecting AssetTargetFallback for package dependencies states: https://github.com/NuGet/Home/blob/dev/proposed/2022/AssetTargetFallback-DependenciesResolution.md#data-analysis---are-packages-authored-consistently

> 3.1% (5) of all (really only 85% can call into this) the packages have inconsistencies in the target frameworks defined in the compile and dependency groups. These packages can be considered not well authored and are specifically affected by the direction we choose. Pack warning NU5128 informs package authors that their package has this inconsistency, and the percentage of packages with this inconsistency should decrease as authors adopt the warning's suggestions.

So, feedback that you're experiencing this issue is valuable, if we get a lot of feedback that customers are having build failures, we can take action. But currently I agree with Nikolche that it's by design and is a symptom of packages that are not well authored. If package authors fixed warnings, NU5128 specifically, this type of package consumer issue would not happen.

Perhaps it was a misunderstanding that missing dependency group != empty dependency group. As I previously explained, NuGet has always selected the "best" dependency group, even when there isn't an exact match based on file assets selected. When the packages used conform to NU5128, package consumers benefit from not having runtime errors now that dependency resolution uses AssetTargetFallback.

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs
@@ -87,13 +87,13 @@ namespace NuGet.Commands
         /// is <c>null</c>.</exception>
         /// <exception cref="ArgumentNullException">Thrown if <paramref name="logger" /> is <c>null</c>.</exception>
         public SourceRepositoryDependencyProvider(
-        SourceRepository sourceRepository,
-        ILogger logger,
-        SourceCacheContext cacheContext,
-        bool ignoreFailedSources,
-        bool ignoreWarning,
-        LocalPackageFileCache fileCache,
-        bool isFallbackFolderSource) :
+            SourceRepository sourceRepository,
+            ILogger logger,
+            SourceCacheContext cacheContext,
+            bool ignoreFailedSources,
+            bool ignoreWarning,
+            LocalPackageFileCache fileCache,
+            bool isFallbackFolderSource) :
             this(sourceRepository,
                 logger,
                 cacheContext,
@@ -106,14 +106,14 @@ namespace NuGet.Commands
         }
 
         internal SourceRepositoryDependencyProvider(
-        SourceRepository sourceRepository,
-        ILogger logger,
-        SourceCacheContext cacheContext,
-        bool ignoreFailedSources,
-        bool ignoreWarning,
-        LocalPackageFileCache fileCache,
-        bool isFallbackFolderSource,
-        IEnvironmentVariableReader environmentVariableReader)
+            SourceRepository sourceRepository,
+            ILogger logger,
+            SourceCacheContext cacheContext,
+            bool ignoreFailedSources,
+            bool ignoreWarning,
+            LocalPackageFileCache fileCache,
+            bool isFallbackFolderSource,
+            IEnvironmentVariableReader environmentVariableReader)
         {
             _sourceRepository = sourceRepository ?? throw new ArgumentNullException(nameof(sourceRepository));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs
@@ -34,6 +34,7 @@ namespace NuGet.Commands
         private bool _ignoreFailedSources;
         private bool _ignoreWarning;
         private bool _isFallbackFolderSource;
+        private bool _useLegacyAssetTargetFallbackBehavior;
 
         private readonly ConcurrentDictionary<LibraryRangeCacheKey, AsyncLazy<LibraryDependencyInfo>> _dependencyInfoCache
             = new ConcurrentDictionary<LibraryRangeCacheKey, AsyncLazy<LibraryDependencyInfo>>();
@@ -92,7 +93,27 @@ namespace NuGet.Commands
         bool ignoreFailedSources,
         bool ignoreWarning,
         LocalPackageFileCache fileCache,
-        bool isFallbackFolderSource)
+        bool isFallbackFolderSource) :
+            this(sourceRepository,
+                logger,
+                cacheContext,
+                ignoreFailedSources,
+                ignoreWarning,
+                fileCache,
+                isFallbackFolderSource,
+                environmentVariableReader: EnvironmentVariableWrapper.Instance)
+        {
+        }
+
+        internal SourceRepositoryDependencyProvider(
+        SourceRepository sourceRepository,
+        ILogger logger,
+        SourceCacheContext cacheContext,
+        bool ignoreFailedSources,
+        bool ignoreWarning,
+        LocalPackageFileCache fileCache,
+        bool isFallbackFolderSource,
+        IEnvironmentVariableReader environmentVariableReader)
         {
             _sourceRepository = sourceRepository ?? throw new ArgumentNullException(nameof(sourceRepository));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
@@ -101,6 +122,7 @@ namespace NuGet.Commands
             _ignoreWarning = ignoreWarning;
             _packageFileCache = fileCache;
             _isFallbackFolderSource = isFallbackFolderSource;
+            _useLegacyAssetTargetFallbackBehavior = MSBuildStringUtility.IsTrue(environmentVariableReader.GetEnvironmentVariable(Environment.GetEnvironmentVariable("NUGET_USE_LEGACY_ASSET_TARGET_FALLBACK_DEPENDENCY_RESOLUTION")));
         }
 
         /// <summary>
@@ -293,6 +315,7 @@ namespace NuGet.Commands
 
             AsyncLazy<LibraryDependencyInfo> result = null;
 
+            // Therre's caching here which happens based on a bunch of stuff.
             var action = new AsyncLazy<LibraryDependencyInfo>(async () =>
                 await GetDependenciesCoreAsync(libraryIdentity, targetFramework, cacheContext, logger, cancellationToken));
 
@@ -470,7 +493,7 @@ namespace NuGet.Commands
             return null;
         }
 
-        private static IEnumerable<LibraryDependency> GetDependencies(
+        private IEnumerable<LibraryDependency> GetDependencies(
             FindPackageByIdDependencyInfo packageInfo,
             NuGetFramework targetFramework)
         {
@@ -488,13 +511,17 @@ namespace NuGet.Commands
                 dependencyGroup = NuGetFrameworkUtility.GetNearest(packageInfo.DependencyGroups, dualCompatibilityFramework.SecondaryFramework, item => item.TargetFramework);
             }
 
-            // FrameworkReducer.GetNearest does not consider ATF since it is used for more than just compat
-            if (dependencyGroup == null &&
-                targetFramework is AssetTargetFallbackFramework assetTargetFallbackFramework)
+            if (!_useLegacyAssetTargetFallbackBehavior)
             {
-                dependencyGroup = NuGetFrameworkUtility.GetNearest(packageInfo.DependencyGroups,
-                    assetTargetFallbackFramework.AsFallbackFramework(),
-                    item => item.TargetFramework);
+                // FrameworkReducer.GetNearest does not consider ATF since it is used for more than just compat
+
+                if (dependencyGroup == null &&
+                    targetFramework is AssetTargetFallbackFramework assetTargetFallbackFramework)
+                {
+                    dependencyGroup = NuGetFrameworkUtility.GetNearest(packageInfo.DependencyGroups,
+                        assetTargetFallbackFramework.AsFallbackFramework(),
+                        item => item.TargetFramework);
+                }
             }
 
             if (dependencyGroup != null)

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs
@@ -315,7 +315,6 @@ namespace NuGet.Commands
 
             AsyncLazy<LibraryDependencyInfo> result = null;
 
-            // Therre's caching here which happens based on a bunch of stuff.
             var action = new AsyncLazy<LibraryDependencyInfo>(async () =>
                 await GetDependenciesCoreAsync(libraryIdentity, targetFramework, cacheContext, logger, cancellationToken));
 

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs
@@ -122,7 +122,7 @@ namespace NuGet.Commands
             _ignoreWarning = ignoreWarning;
             _packageFileCache = fileCache;
             _isFallbackFolderSource = isFallbackFolderSource;
-            _useLegacyAssetTargetFallbackBehavior = MSBuildStringUtility.IsTrue(environmentVariableReader.GetEnvironmentVariable(Environment.GetEnvironmentVariable("NUGET_USE_LEGACY_ASSET_TARGET_FALLBACK_DEPENDENCY_RESOLUTION")));
+            _useLegacyAssetTargetFallbackBehavior = MSBuildStringUtility.IsTrue(environmentVariableReader.GetEnvironmentVariable("NUGET_USE_LEGACY_ASSET_TARGET_FALLBACK_DEPENDENCY_RESOLUTION"));
         }
 
         /// <summary>

--- a/src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs
+++ b/src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs
@@ -27,9 +27,21 @@ namespace NuGet.ProjectModel
 
         private readonly ILogger _logger;
 
+        private readonly bool _useLegacyAssetTargetFallbackBehavior;
+
         public PackageSpecReferenceDependencyProvider(
             IEnumerable<ExternalProjectReference> externalProjects,
-            ILogger logger)
+            ILogger logger) :
+            this(externalProjects,
+                logger,
+                environmentVariableReader: EnvironmentVariableWrapper.Instance)
+        {
+        }
+
+        internal PackageSpecReferenceDependencyProvider(
+            IEnumerable<ExternalProjectReference> externalProjects,
+            ILogger logger,
+            IEnvironmentVariableReader environmentVariableReader)
         {
             if (externalProjects == null)
             {
@@ -68,6 +80,7 @@ namespace NuGet.ProjectModel
                     _externalProjectsByUniqueName.Add(project.UniqueName, project);
                 }
             }
+            _useLegacyAssetTargetFallbackBehavior = MSBuildStringUtility.IsTrue(environmentVariableReader.GetEnvironmentVariable("NUGET_USE_LEGACY_ASSET_TARGET_FALLBACK_DEPENDENCY_RESOLUTION"));
         }
 
         public bool SupportsType(LibraryDependencyTarget libraryType)
@@ -155,7 +168,7 @@ namespace NuGet.ProjectModel
             return library;
         }
 
-        private static void AddLibraryProperties(Library library, PackageSpec packageSpec, NuGetFramework targetFramework)
+        private void AddLibraryProperties(Library library, PackageSpec packageSpec, NuGetFramework targetFramework)
         {
             var projectStyle = packageSpec.RestoreMetadata?.ProjectStyle ?? ProjectStyle.Unknown;
 
@@ -186,9 +199,12 @@ namespace NuGet.ProjectModel
                 var targetFrameworkInfo = packageSpec.GetTargetFramework(targetFramework);
 
                 // FrameworkReducer.GetNearest does not consider ATF since it is used for more than just compat
-                if (targetFrameworkInfo.FrameworkName == null && targetFramework is AssetTargetFallbackFramework atfFramework)
+                if (!_useLegacyAssetTargetFallbackBehavior)
                 {
-                    targetFrameworkInfo = packageSpec.GetTargetFramework(atfFramework.AsFallbackFramework());
+                    if (targetFrameworkInfo.FrameworkName == null && targetFramework is AssetTargetFallbackFramework atfFramework)
+                    {
+                        targetFrameworkInfo = packageSpec.GetTargetFramework(atfFramework.AsFallbackFramework());
+                    }
                 }
 
                 if (targetFrameworkInfo.FrameworkName == null && targetFramework is DualCompatibilityFramework mcfFramework)
@@ -219,10 +235,13 @@ namespace NuGet.ProjectModel
             // Get the nearest framework
             var referencesForFramework = packageSpec.GetRestoreMetadataFramework(targetFramework);
 
-            if (referencesForFramework.FrameworkName == null &&
-                  targetFramework is AssetTargetFallbackFramework assetTargetFallbackFramework)
+            if (!_useLegacyAssetTargetFallbackBehavior)
             {
-                referencesForFramework = packageSpec.GetRestoreMetadataFramework(assetTargetFallbackFramework.AsFallbackFramework());
+                if (referencesForFramework.FrameworkName == null &&
+                      targetFramework is AssetTargetFallbackFramework assetTargetFallbackFramework)
+                {
+                    referencesForFramework = packageSpec.GetRestoreMetadataFramework(assetTargetFallbackFramework.AsFallbackFramework());
+                }
             }
 
             // Ensure that this project is compatible
@@ -330,6 +349,7 @@ namespace NuGet.ProjectModel
                 dependencies.AddRange(packageSpec.Dependencies);
 
                 // Add framework specific dependencies
+                // Disable?
                 var targetFrameworkInfo = packageSpec.GetTargetFramework(targetFramework);
 
                 if (targetFrameworkInfo.FrameworkName == null && targetFramework is AssetTargetFallbackFramework atfFramework)

--- a/src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs
+++ b/src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs
@@ -199,12 +199,9 @@ namespace NuGet.ProjectModel
                 var targetFrameworkInfo = packageSpec.GetTargetFramework(targetFramework);
 
                 // FrameworkReducer.GetNearest does not consider ATF since it is used for more than just compat
-                if (!_useLegacyAssetTargetFallbackBehavior)
+                if (targetFrameworkInfo.FrameworkName == null && targetFramework is AssetTargetFallbackFramework atfFramework)
                 {
-                    if (targetFrameworkInfo.FrameworkName == null && targetFramework is AssetTargetFallbackFramework atfFramework)
-                    {
-                        targetFrameworkInfo = packageSpec.GetTargetFramework(atfFramework.AsFallbackFramework());
-                    }
+                    targetFrameworkInfo = packageSpec.GetTargetFramework(atfFramework.AsFallbackFramework());
                 }
 
                 if (targetFrameworkInfo.FrameworkName == null && targetFramework is DualCompatibilityFramework mcfFramework)
@@ -337,7 +334,7 @@ namespace NuGet.ProjectModel
             return dependencies;
         }
 
-        internal static List<LibraryDependency> GetSpecDependencies(
+        internal List<LibraryDependency> GetSpecDependencies(
             PackageSpec packageSpec,
             NuGetFramework targetFramework)
         {
@@ -349,12 +346,14 @@ namespace NuGet.ProjectModel
                 dependencies.AddRange(packageSpec.Dependencies);
 
                 // Add framework specific dependencies
-                // Disable?
                 var targetFrameworkInfo = packageSpec.GetTargetFramework(targetFramework);
 
-                if (targetFrameworkInfo.FrameworkName == null && targetFramework is AssetTargetFallbackFramework atfFramework)
+                if (!_useLegacyAssetTargetFallbackBehavior)
                 {
-                    targetFrameworkInfo = packageSpec.GetTargetFramework(atfFramework.AsFallbackFramework());
+                    if (targetFrameworkInfo.FrameworkName == null && targetFramework is AssetTargetFallbackFramework atfFramework)
+                    {
+                        targetFrameworkInfo = packageSpec.GetTargetFramework(atfFramework.AsFallbackFramework());
+                    }
                 }
 
                 dependencies.AddRange(targetFrameworkInfo.Dependencies);

--- a/test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs
@@ -808,11 +808,6 @@ namespace NuGet.Commands.Test
             source.SetupGet(s => s.PackageSource)
                 .Returns(new PackageSource("http://test/index.json"));
 
-            var libraryRange = new LibraryRange(
-                "x",
-                new VersionRange(new NuGetVersion(1, 0, 0, "beta")),
-                LibraryDependencyTarget.Package);
-
             var provider = new SourceRepositoryDependencyProvider(
                 source.Object,
                 testLogger,
@@ -874,11 +869,6 @@ namespace NuGet.Commands.Test
                 .ReturnsAsync(findResource.Object);
             source.SetupGet(s => s.PackageSource)
                 .Returns(new PackageSource("http://test/index.json"));
-
-            var libraryRange = new LibraryRange(
-                "x",
-                new VersionRange(new NuGetVersion(1, 0, 0, "beta")),
-                LibraryDependencyTarget.Package);
 
             var provider = new SourceRepositoryDependencyProvider(
                 source.Object,

--- a/test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Commands.Test/SourceRepositoryDependencyProviderTests.cs
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
@@ -17,6 +18,7 @@ using NuGet.Protocol;
 using NuGet.Protocol.Core.Types;
 using NuGet.Test.Utility;
 using NuGet.Versioning;
+using Test.Utility;
 using Xunit;
 
 namespace NuGet.Commands.Test
@@ -770,6 +772,141 @@ namespace NuGet.Commands.Test
             results.Should().Be(null);
             secondTestLogger.WarningMessages.Should().HaveCount(1);
             secondTestLogger.ShowWarnings().Should().Contain("NU1801");
+        }
+
+        [Fact]
+        public async Task GetDependenciesAsync_WhenPackageHasVariousDependencyGroups_CorrectFrameworkIsSelected()
+        {
+            // Arrange
+            var testLogger = new TestLogger();
+            var cacheContext = new SourceCacheContext();
+            var findResource = new Mock<FindPackageByIdResource>();
+            var net46 = FrameworkConstants.CommonFrameworks.Net46;
+            var netstandard20 = FrameworkConstants.CommonFrameworks.NetStandard20;
+
+            var packageDependencyGroups = new List<PackageDependencyGroup>();
+            var net46group = new PackageDependencyGroup(net46, new PackageDependency[] { new PackageDependency("full.framework", VersionRange.Parse("1.0.0")) });
+            var netstandardGroup = new PackageDependencyGroup(netstandard20, new PackageDependency[] { new PackageDependency("netstandard", VersionRange.Parse("1.0.0")) });
+
+            packageDependencyGroups.Add(net46group);
+            packageDependencyGroups.Add(netstandardGroup);
+
+            findResource.Setup(s => s.GetDependencyInfoAsync(
+                    It.IsAny<string>(),
+                    It.IsAny<NuGetVersion>(),
+                    It.IsAny<SourceCacheContext>(),
+                    It.IsAny<ILogger>(),
+                    It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new FindPackageByIdDependencyInfo(
+                    new PackageIdentity("x", NuGetVersion.Parse("1.0.0-beta")),
+                    packageDependencyGroups,
+                    Enumerable.Empty<FrameworkSpecificGroup>()));
+
+            var source = new Mock<SourceRepository>();
+            source.Setup(s => s.GetResourceAsync<FindPackageByIdResource>())
+                .ReturnsAsync(findResource.Object);
+            source.SetupGet(s => s.PackageSource)
+                .Returns(new PackageSource("http://test/index.json"));
+
+            var libraryRange = new LibraryRange(
+                "x",
+                new VersionRange(new NuGetVersion(1, 0, 0, "beta")),
+                LibraryDependencyTarget.Package);
+
+            var provider = new SourceRepositoryDependencyProvider(
+                source.Object,
+                testLogger,
+                cacheContext,
+                ignoreFailedSources: true,
+                ignoreWarning: true,
+                fileCache: null,
+                isFallbackFolderSource: false,
+                new TestEnvironmentVariableReader(new Dictionary<string, string>()));
+
+            // Act
+            var library = await provider.GetDependenciesAsync(
+                new LibraryIdentity("x", NuGetVersion.Parse("1.0.0-beta"), LibraryType.Package),
+                FrameworkConstants.CommonFrameworks.Net472,
+                cacheContext,
+                testLogger,
+                CancellationToken.None);
+            // Assert
+            library.Dependencies.Should().HaveCount(1);
+            var dependencies = library.Dependencies.Single();
+            dependencies.Name.Should().Be("full.framework");
+        }
+
+        [Theory]
+        [InlineData("true", false)]
+        [InlineData("blbla", true)]
+        public async Task GetDependenciesAsync_WhenPackageIsSelectedWithAssetTargetFallback_AndLegacyDependencyResolutionVariableIsSpecified_CorrectDependenciesAreSelected(string envValue, bool areDependenciesSelected)
+        {
+            // Arrange
+            var testLogger = new TestLogger();
+            var cacheContext = new SourceCacheContext();
+            var findResource = new Mock<FindPackageByIdResource>();
+            var wrapper = new TestEnvironmentVariableReader(new Dictionary<string, string>
+            {
+                { "NUGET_USE_LEGACY_ASSET_TARGET_FALLBACK_DEPENDENCY_RESOLUTION", envValue }
+            });
+            var net472 = FrameworkConstants.CommonFrameworks.Net472;
+            var net60 = FrameworkConstants.CommonFrameworks.Net60;
+            var inputFramework = new AssetTargetFallbackFramework(net60, new List<NuGetFramework> { net472 });
+
+            var packageDependencyGroups = new List<PackageDependencyGroup>();
+            var net472Group = new PackageDependencyGroup(net472, new PackageDependency[] { new PackageDependency("full.framework", VersionRange.Parse("1.0.0")) });
+
+            packageDependencyGroups.Add(net472Group);
+
+            findResource.Setup(s => s.GetDependencyInfoAsync(
+                    It.IsAny<string>(),
+                    It.IsAny<NuGetVersion>(),
+                    It.IsAny<SourceCacheContext>(),
+                    It.IsAny<ILogger>(),
+                    It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new FindPackageByIdDependencyInfo(
+                    new PackageIdentity("x", NuGetVersion.Parse("1.0.0-beta")),
+                    packageDependencyGroups,
+                    Enumerable.Empty<FrameworkSpecificGroup>()));
+
+            var source = new Mock<SourceRepository>();
+            source.Setup(s => s.GetResourceAsync<FindPackageByIdResource>())
+                .ReturnsAsync(findResource.Object);
+            source.SetupGet(s => s.PackageSource)
+                .Returns(new PackageSource("http://test/index.json"));
+
+            var libraryRange = new LibraryRange(
+                "x",
+                new VersionRange(new NuGetVersion(1, 0, 0, "beta")),
+                LibraryDependencyTarget.Package);
+
+            var provider = new SourceRepositoryDependencyProvider(
+                source.Object,
+                testLogger,
+                cacheContext,
+                ignoreFailedSources: true,
+                ignoreWarning: true,
+                fileCache: null,
+                isFallbackFolderSource: false,
+                wrapper);
+
+            // Act
+            var library = await provider.GetDependenciesAsync(
+                new LibraryIdentity("x", NuGetVersion.Parse("1.0.0-beta"), LibraryType.Package),
+                inputFramework,
+                cacheContext,
+                testLogger,
+                CancellationToken.None);
+            // Assert
+            if (areDependenciesSelected)
+            {
+                library.Dependencies.Should().HaveCount(1);
+                var dependencies = library.Dependencies.Single();
+                dependencies.Name.Should().Be("full.framework");
+            } else
+            {
+                library.Dependencies.Should().HaveCount(0);
+            }
         }
 
         private sealed class SourceRepositoryDependencyProviderTest : IDisposable

--- a/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecReferenceDependencyProviderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecReferenceDependencyProviderTests.cs
@@ -3,10 +3,13 @@
 
 using System.Collections.Generic;
 using System.Linq;
+using FluentAssertions;
+using NuGet.Commands.Test;
+using NuGet.Common;
 using NuGet.Frameworks;
 using NuGet.LibraryModel;
-using NuGet.Test.Utility;
 using NuGet.Versioning;
+using Test.Utility;
 using Xunit;
 
 namespace NuGet.ProjectModel.Test
@@ -40,8 +43,9 @@ namespace NuGet.ProjectModel.Test
             var dependencyGraphSpec = CreateDependencyGraphSpecWithCentralDependencies(cpvmEnabled, CentralPackageTransitivePinningEnabled, tfi);
             var packSpec = dependencyGraphSpec.Projects[0];
 
+            var dependencyProvider = new PackageSpecReferenceDependencyProvider(new List<ExternalProjectReference>(), NullLogger.Instance);
             // Act
-            var dependencies = PackageSpecReferenceDependencyProvider.GetSpecDependencies(packSpec, tfi.FrameworkName);
+            var dependencies = dependencyProvider.GetSpecDependencies(packSpec, tfi.FrameworkName);
 
             // Assert
             if (cpvmEnabled && CentralPackageTransitivePinningEnabled)
@@ -70,6 +74,44 @@ namespace NuGet.ProjectModel.Test
                 Assert.Equal(fooDep.LibraryRange.VersionRange != null, cpvmEnabled);
                 Assert.Equal(LibraryDependencyReferenceType.Direct, fooDep.ReferenceType);
             }
+        }
+
+        [Fact]
+        public void GetSpecDependencies_WithAssetTargetFallback_ReturnsCorrectDependencies()
+        {
+            // Arrange
+            var net60Framework = FrameworkConstants.CommonFrameworks.Net60;
+            var net472Framework = FrameworkConstants.CommonFrameworks.Net472;
+            var packageSpec = ProjectTestHelpers.GetPackageSpec(rootPath: "C:\\", projectName: "A", framework: net472Framework.GetShortFolderName(), dependencyName: "x");
+
+            var dependencyProvider = new PackageSpecReferenceDependencyProvider(new List<ExternalProjectReference>(), NullLogger.Instance);
+            var assetTargetFallback = new AssetTargetFallbackFramework(net60Framework, new List<NuGetFramework> { net472Framework });
+            // Act
+
+            var dependencies = dependencyProvider.GetSpecDependencies(packageSpec, assetTargetFallback);
+
+            // Assert
+            dependencies.Should().HaveCount(1);
+            dependencies[0].Name.Should().Be("x");
+        }
+
+        [Fact]
+        public void GetSpecDependencies_WithAssetTargetFallback_AndDependencyResolutionDisabled_ReturnsCorrectDependencies()
+        {
+            // Arrange
+            var net60Framework = FrameworkConstants.CommonFrameworks.Net60;
+            var net472Framework = FrameworkConstants.CommonFrameworks.Net472;
+            var packageSpec = ProjectTestHelpers.GetPackageSpec(rootPath: "C:\\", projectName: "A", framework: net472Framework.GetShortFolderName(), dependencyName: "x");
+
+            var envVarWrapper = new TestEnvironmentVariableReader(new Dictionary<string, string> { { "NUGET_USE_LEGACY_ASSET_TARGET_FALLBACK_DEPENDENCY_RESOLUTION", "true" } });
+            var dependencyProvider = new PackageSpecReferenceDependencyProvider(new List<ExternalProjectReference>(), NullLogger.Instance, envVarWrapper);
+            var assetTargetFallback = new AssetTargetFallbackFramework(net60Framework, new List<NuGetFramework> { net472Framework });
+            // Act
+
+            var dependencies = dependencyProvider.GetSpecDependencies(packageSpec, assetTargetFallback);
+
+            // Assert
+            dependencies.Should().HaveCount(0);
         }
 
         private static TargetFrameworkInformation CreateTargetFrameworkInformation(List<LibraryDependency> dependencies, List<CentralPackageVersion> centralVersionsDependencies, bool cpvmEnabled)

--- a/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecReferenceDependencyProviderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecReferenceDependencyProviderTests.cs
@@ -76,34 +76,17 @@ namespace NuGet.ProjectModel.Test
             }
         }
 
-        [Fact]
-        public void GetSpecDependencies_WithAssetTargetFallback_ReturnsCorrectDependencies()
+        [Theory]
+        [InlineData(null, 1)]
+        [InlineData("true", 0)]
+        public void GetSpecDependencies_WithAssetTargetFallback_AndDependencyResolutionVariableSpecified_ReturnsCorrectDependencies(string assetTargetFallbackEnvironmentVariableValue, int dependencyCount)
         {
             // Arrange
             var net60Framework = FrameworkConstants.CommonFrameworks.Net60;
             var net472Framework = FrameworkConstants.CommonFrameworks.Net472;
             var packageSpec = ProjectTestHelpers.GetPackageSpec(rootPath: "C:\\", projectName: "A", framework: net472Framework.GetShortFolderName(), dependencyName: "x");
 
-            var dependencyProvider = new PackageSpecReferenceDependencyProvider(new List<ExternalProjectReference>(), NullLogger.Instance);
-            var assetTargetFallback = new AssetTargetFallbackFramework(net60Framework, new List<NuGetFramework> { net472Framework });
-            // Act
-
-            var dependencies = dependencyProvider.GetSpecDependencies(packageSpec, assetTargetFallback);
-
-            // Assert
-            dependencies.Should().HaveCount(1);
-            dependencies[0].Name.Should().Be("x");
-        }
-
-        [Fact]
-        public void GetSpecDependencies_WithAssetTargetFallback_AndDependencyResolutionDisabled_ReturnsCorrectDependencies()
-        {
-            // Arrange
-            var net60Framework = FrameworkConstants.CommonFrameworks.Net60;
-            var net472Framework = FrameworkConstants.CommonFrameworks.Net472;
-            var packageSpec = ProjectTestHelpers.GetPackageSpec(rootPath: "C:\\", projectName: "A", framework: net472Framework.GetShortFolderName(), dependencyName: "x");
-
-            var envVarWrapper = new TestEnvironmentVariableReader(new Dictionary<string, string> { { "NUGET_USE_LEGACY_ASSET_TARGET_FALLBACK_DEPENDENCY_RESOLUTION", "true" } });
+            var envVarWrapper = new TestEnvironmentVariableReader(new Dictionary<string, string> { { "NUGET_USE_LEGACY_ASSET_TARGET_FALLBACK_DEPENDENCY_RESOLUTION", assetTargetFallbackEnvironmentVariableValue } });
             var dependencyProvider = new PackageSpecReferenceDependencyProvider(new List<ExternalProjectReference>(), NullLogger.Instance, envVarWrapper);
             var assetTargetFallback = new AssetTargetFallbackFramework(net60Framework, new List<NuGetFramework> { net472Framework });
             // Act
@@ -111,7 +94,12 @@ namespace NuGet.ProjectModel.Test
             var dependencies = dependencyProvider.GetSpecDependencies(packageSpec, assetTargetFallback);
 
             // Assert
-            dependencies.Should().HaveCount(0);
+            dependencies.Should().HaveCount(dependencyCount);
+
+            if (dependencyCount > 0)
+            {
+                dependencies[0].Name.Should().Be("x");
+            }
         }
 
         private static TargetFrameworkInformation CreateTargetFrameworkInformation(List<LibraryDependency> dependencies, List<CentralPackageVersion> centralVersionsDependencies, bool cpvmEnabled)

